### ISSUE TITLE
upsert is not possible on MongoDB

### ISSUE THREAD

**Issue by ramiel** [prisma/prisma#20157]
### Bug description

Creating an upsert on mongoDB is not possible because an id is always required. Let's consider this case

```ts
return this.prisma.entity.upsert({
      where: { id: id },
      create: createData,
      update: updateData,
    });
```
In case we need a creation, there's no good value of `id` that can fire it.     
If it is `undefined` it won't work because id must have a value.    
If it is an empty string, it cannot work because it must be an ObjectId.

Creating an objectId client side just to prove there's no one exiting already, looks incorrect to me, or at least a bad design decision.

This is related to issue #5233

### How to reproduce

Run the code above with a MongoDB datasource

### Expected behavior

In the other mentioned issue there are some proposed solution. My favorite is to allow id to be `undefined` but this has some detractors whose arguments I don't fully understand/agree with.

### Prisma information

Not relevant. anything where the id is

```
model Entity {
id             String          @id @default(auto()) @map("_id") @db.ObjectId
}
```

### Environment & setup

- OS: any
- Database: MongoDB
- Node.js version: 18

### Prisma Version

```
4.16.2
```

**Comment by millsp** [comment]
Thanks for your report @ramiel. Yes, we should definitely improve our API to support this use-case.

> Creating an objectId client side just to prove there's no one exiting already, looks incorrect to me, or at least a bad design decision.

For now that is the workaround, but I fully agree with you.

**Comment by ramiel** [comment]
I see. The only problem with this solution is that it triggers a find on the DB which is useless. For the moment is better NOT to use the upsert method at all.

 Looking forward for any improvement on this and let me know if I can help somehow.

**Comment by saalihou** [comment]
The only workaround I can think of is to implement a unique key constraints, catch it during insert and then do a find, which is a bit tedious and worse for performance.

But I'm wondering day to day if anyone is using Prisma with MongoDB in production. Because missing an actual `upsert` operation seems like a big deal for an ORM.

**Comment by janpio** [comment]
@ramiel Can you expand a bit more what situation exactly is not possible right now? Your code example implies you have an `id` already you want to use for the `upsert()`. Am I misunderstanding?

**Comment by ramiel** [comment]
No, the id can be given or `undefined`. As for my understanding this problem is better explained in https://github.com/prisma/prisma/issues/5233 and for some reason there's no will to solve it, or at the least the prisma team can't agree on a proper solution

**Comment by janpio** [comment]
I am part of the Prisma team, trying to understand 😆 

So this problem is not actually MongoDB specific, but is about the higher level `undefined` behavior with `upsert` that is discussed in #5233? Or is there soem MongoDB specific angle that justifies this being tracked separately and with MongoDB specific description?

**Comment by ramiel** [comment]
While the problem is common for all DB, in mongodb is even worse because as id you can only provide an ObjectId (in sql is usually an number or a string), so the solution to provide a fake id is more difficult because the user must generate a valid objectId like value since it's checked. That said, this issue can be probably ignored if a global solution is found. Also, about my "the prisma team can't agree on a proper solution" is not a criticism, but it's what you can tell from the other issue replies. Happy to see interest in this! :)

**Comment by janpio** [comment]
Ok, I understand that just writing `0` for an `Int` field (that is expected to only contain positive numbers anyway) is a lot easier than crafting a valid ObjectId.

But that also means that this issue indeed is just a special case of #5233, and we can probably close this and make sure to highlight the MongoDB specific problems over there as a comment. I'll take care of that.

**Comment by ramiel** [comment]
Theanks @janpio

**Comment by janpio** [comment]
Why do you want to use an `upsert()` though when you already know that the object does not exist in the database? Why not just give it to `create()`?

**Comment by ramiel** [comment]
Not that simple. The id (that can be undefined) comes usually from user interaction. So the code should be something like

```
if(id){
  prisma.update()
} else {
  prisma.create
}
```
but this is exactly the use case of an upsert. I mean, why do you need the `upsert` command at all when you know if you want to udpate or create?
